### ISSUE TITLE
Support background color erase (bce) [was: Vim copy & paste trailing space issue]

### ISSUE THREAD

**Issue by mikehaertl** [tmux/tmux#109]
When I run Vim from inside a tmux session and copy some text to the clipboard, each line gets padded with spaces to the right.

For example, say I have a text file like this:

```
^some$
^text$
```

**Note**: I've used `^` and `$` to mark the beginning and end of a line respectively. They are *not* part of the file content.

I start tmux and open this file in Vim. I press <kbd>Shift</kbd> (to prevent Vim from processing the mouse click) and mark the complete text by holding the left mouse button clicked. Then I copy it to the clipboard with <kbd>Shift</kbd>+<kbd>Ctrl</kbd>+<kbd>c</kbd>.

The result in the clipboard is something like:

```
^some                                                  $
^text                                                  $
```

Note the extra spaces. (The number of spaces depends on the width of the terminal window.)

If I start Vim without tmux or if I just `cat` the file content and then copy it, there are no extra spaces. So it must have to do with the combination of Vim + tmux. I've seen this on different Linux flavours, i.e. on Ubuntu and Mint. I use the default terminal (Gnome Terminal 3.6.2) there.

So how can I prevent this?

My `tmux.conf`

```
set-option -g prefix C-a
set-option -g mouse-utf8 off
set-option -g status-keys vi
set-window-option -g mode-keys vi
set-window-option -g mode-mouse on
set-option -g terminal-overrides 'xterm*:smcup@:rmcup@'

bind-key C-a last-window
bind-key C-h select-pane -L
bind-key C-l select-pane -R

# colors:
set-option -g status-bg black
set-option -g status-fg white
set-option -g status-left '#[fg=green]#H'
set-window-option -g window-status-current-bg red
```

I've also tried without the above `.tmux.conf`, using tmux' default settings - no difference.

Here's my `/etc/vim/vimrc.local` that I use on Mint (Ubuntu derivative):
https://gist.github.com/mikehaertl/1612035

**Comment by mikehaertl** [comment]
One more note: The only workaround so far was, to set `default-terminal` to `xterm` - but then the background colors are messed up in Vim. They then end at the end of each line.

**Comment by nicm** [comment]
You can set trimSelection in xterm. Or try to figure out what is different in the xterm terminfo entry that makes vim erase lines properly rather than using spaces. Although there is no guarantee tmux itself wouldn't use spaces if it was better to do so.

**Comment by mikehaertl** [comment]
Hmm, maybe I misunderstood, but I use Gnome Terminal, not xterm. If that's not what you meant, then where would I make this adjustment? And any hints how i can investigate the differences between `xterm` and `screen` terminfo and what to do if I find any?

**Comment by mikehaertl** [comment]
Ok, I did some googling and found [this](http://comments.gmane.org/gmane.mail.mutt.user/35997) somewhat related discussion. Quote:

> I'm pretty sure this has to do with the use of bkgdset() so that clrtoeol() will set the proper background color all the way to the end of the line. Consider the case where you have a different default background terminal color than what the user has specified in 'color body ...'.  Without the use of bkgdset(), you'd see the wrong color at the end of each line.

That's pretty much what happens if i set `xterm` in tmux`s `default-terminal`. Another comment says:

> If you don't use bkgdset(), any time there is a non-default background color used and the text does not reach the right column, the space from the end of the text to the right column will be the background color rather than the color the user requested.
>>About that problem, the FAQ tells to use the screen's bce option, but that
>>does not work really well…
> I put this in my ~/.screenrc:
>
>	defbce on
>	term screen-bce

I've checked with `infocmp xterm | grep bce` and `infocmp screen | grep bce`. The `bce` option is really missing in `screen`. So I've tried again with `set-option -g default-terminal "screen-bce"` in my tmux config. Unfortunately the background color to the right of the line end is still messed up in Vim with this `TERM` setting. 

It feels like getting closer to a solution, though. Any further ideas what I could try?

**Comment by mikehaertl** [comment]
To met it seems like the key part is the `defbce on` setting that screen has. Is there a similar option in tmux?

**Comment by mikehaertl** [comment]
I've also tried this setting in Vim:

```
set t_ut=
```

This turns of backround color erase in Vim. It fixes the Bg color issues - but now copy & past has the trailing spaces again. I'm turning in circles.

I still feel, like this needs to be fixed in tmux somehow.

**Comment by mikehaertl** [comment]
Ok, after some more investigation I think i can refine my issue as a feature request for background color erase. The original `screen` command hat support for it, tmux seems to miss it.

> The term "bce" stands for "back color erase". Terminals such as modern xterm and rxvt implement back color erase, others such as dtterm do not. (Roughly half of the emulators that I know about implement bce). When an application clears the screen, a terminal that implements back color erase will retain the last-set background color. A terminal that does not implement back color erase will reset the background color to the default or initial colors. Applications that paint most of the screen in a single color are more efficient on terminals that support back color erase. Inevitably, there are tradeoffs and issues with standardization of the feature as noted in the ncurses FAQ. Unsurprisingly, ncurses supports xterm's behavior.

**Comment by nicm** [comment]
tmux doesn't support BCE.

**Comment by nicm** [comment]
Your best bet is to see if gnome terminal has an option to trim spaces when copying the selection like xterm.

**Comment by mikehaertl** [comment]
Hmm, ok. Just curious: Is this so hard to implement? Or what is the reason that this can't be added to tmux as a feature?

**Comment by nicm** [comment]
Nobody has implemented it.

**Comment by mikehaertl** [comment]
Oh, so does that mean, that tmux is no longer under active development? I wasn't aware of that.

**Comment by nicm** [comment]
No, it means nobody has implemented BCE. I don't care about BCE and so far nobody else has either.

**Comment by mikehaertl** [comment]
Ok, that was loud and clear. Thanks anyway.

**Comment by nicm** [comment]
I thought we'd have to do it recently to do pane background colours, but it wasn't required and the person who it just did enough for pane background colours and didn't take it further :-). Now that that's done we have to support drawing backgrounds on terminals with and without BCE for pane backgrounds anyway, so tmux itself supporting BCE is just a matter of changing the way that all the erase escape sequences work which is not a huge amout of work (but not trivial either).

**Comment by mikehaertl** [comment]
Thanks. Well - that at least gives me some hope. No chance to at least have this feature request added to some backlog? Maybe someone is bored at some point or needs a quick challenge - or just wants to make a single poor soul that suffers from an annoying problem happy again... :P

**Comment by nicm** [comment]
It is probably already in the TODO file, but if not I will add it.

**Comment by mikehaertl** [comment]
Thanks.

**Comment by egmontkob** [comment]
Please see also https://bugzilla.gnome.org/show_bug.cgi?id=754596

**Comment by JoshuaD84** [comment]
I'd be willing to work on a patch for this issue if you could point me in the rough direction of where to start. I'd like to see this implemented.

**Comment by nicm** [comment]
When an application uses one of the clear escape sequences, the code that updates the internal screen (screen-write.c) will need to clear using the background colour, and tmux will need to clear using the background colour when it writes to the terminal (tty.c). Also you may need to make changes to how tmux redraws the screen if it needs to redraw it entirely (screen-redraw.c).

For example for clear to end of line look at screen_write_clearendofline and tty_cmd_clearendofline. There is also clear line, clear start of line, the various clear screen, and scrolling.

tty.c already supports drawing using a nondefault background colour (on both BCE and non-BCE terminals). It'll probably just need minor changes to use the current bg colour instead of the default.

**Comment by 00gavin** [comment]
I've been suffering from the same trailing white space issue here, when selecting multi-line text, using gnome-terminal + ssh + tmux + vim.  I'll be trying: set -g default-terminal "xterm-256color"
This was marked as closed.  Did a patch make it into the latest version?  Cheers.

**Comment by sphaugh** [comment]
@nicm, is there any hope of having this feature? There's a note in the TODO that it's half-completed, could you perhaps give another clue as to what effort is required?

**Comment by ktaragorn** [comment]
I faced this issue for the longest time as well, and setting `default-terminal` as `xterm` helped.

**Comment by corbolais** [comment]
I notice this is closed, no fix available yet? Nobody cared so far?

**Comment by 00gavin** [comment]
Still care. Is this fixed in latest version?

**Comment by corbolais** [comment]
On 2019-03-11 15:08:59, Gavin Burris wrote:
>Still care. Is this fixed in latest version?
not fixed in
- ubuntu 18.10 tmux 2.7-1
- debian tmux 2.8-3

>

**Comment by egmontkob** [comment]
screen/tmux traditionally don't implement bce.

As far as I can see, bce would provide three advantages, only the first one being clearly a real win:

- Better copying behavior with trailing spaces; but this can only work as long as the pane is at the right side of the available area. If it's let's say the left pane of a vertical split then you can't have the desired copy-paste experience.

- Smaller network traffic; honestly does anyone care?

- Less broken behavior for those who incorrectly set TERM=xterm-alike inside tmux, thus fewer support requests or invalid bugreports.

On the other hand, as per the issue I linked, bce is a truly problematic badly designed "feature" in other aspects. E.g. using its traditional interpretation, as seen in xterm, it is literally impossible for a simple utility that doesn't handle the entire screen (such as `ls`) to print a word with a nondefault background color, without risking side effects.

IMHO these altogether might be a strong enough reason for tmux developers not wanting to add this feature.

**Comment by nicm** [comment]
tmux has supported BCE since 2.5 in May 2017.

**Comment by corbolais** [comment]
@egmontkob Thank you for this insight.

So, my desire for bce being implemented in tmux was driven by the fact that the SOA renders copy-and-paste within tmux/vim unusable.

All I care is CnP in tmux/vim, so my point really is: **is there any solution and if so, what is it**. 

I could not find one yet. It seems I'm not the only one struggling at this point.

@nicm thanks for the feedback. How comes then that tmux/vim are still not enabling CnP in a usable way using said distros/sw versions?

Cheers

**Comment by nicm** [comment]
tmux itself doesn't use BCE to draw to the terminal even if the terminal supports it. There might be some small advantages to doing so but it is probably not worth the hassle.

The only way to copy from inside tmux that works no matter how many panes or how tmux chooses to draw to the terminal is to use copy mode and this is what I recommend.

**Comment by mikehaertl** [comment]
> The only way to copy from inside tmux that works no matter how many panes or how tmux chooses to draw to the terminal is to use copy mode and this is what I recommend.

I've not used this before and just tried to figure out how it works. I found some instructions that I should do this:

* Hit <kbd>Ctrl</kbd> + <kbd>b</kbd> followed by <kbd>[</kbd>
* Use the cursor keys to navigate to the beginning of the text
* Hit <kbd>Ctrl</kbd> + <kbd>Space</kbd>
* Move to the end of the selection
* Hit <kbd>Alt</kbd> + <kbd>w</kbd>

Then I can paste the text with <kbd>Ctrl</kbd> + <kbd>b</kbd> <kbd>b</kbd>

I have a few problems with this:

1. ~~It doesn't work. <kbd>Ctrl</kbd> + <kbd>Space</kbd> does nothing for me~~ Correction: It works, but with Vi-style bindings you need to use <kbd>Space</kbd> and <kbd>Enter</kbd> to start/end a selection. The other two points are still valid, though.
2. To me that's a completely awkward flow of keyboard strokes to remember.
3. If it would work, it would still only work inside the tmux session. My standard problem is, that I want to copy and paste text from my Vim session into a browser window. So I need to copy text into the system clipboard.

So unless there's a better workaround I'd still prefer a implementation of BCE in tmux. Couldn't this be a feature, defaulting to `off` and maybe with a big warning that you're on your own if you enable it and have problems?

**Comment by nicm** [comment]
tmux could use BCE to clear areas but it is unlikely I will do it - it
would not necessarily fix the problem and I'm not going to start promising
that tmux will not use spaces if appropriate.

If you want to try to do it I will review it (I suggest you start with
tty_clear_* in tty.c), but it is unlikely to ever be more than best effort
and will never work with panes. Also it can't make tmux slower.

Copy mode is the recommended way to copy from tmux. It works with panes, it
works with the mouse, it works with the X clipboard if your terminal
supports that. If you don't like the key bindings then you can change them.

**Comment by mikehaertl** [comment]
Ok, I've tried with the mouse: I can start copy mode with <kbd>Ctrl</kbd>+<kbd>b</kbd> <kbd>[</kbd>. I then can select text as usual with the mouse, holding left mouse button clicked. When I release the button, the text is no longer highlightet and the selection was not copied into my system clipboard.

Do you maybe have some pointer how to make this work? I'm on Linux Mint, using the default Gnome terminal 3.28.1.

**Comment by nicm** [comment]
Oh I was wrong, tmux will already use BCE to clear when appropriate,
possibly it does not do so on your terminal or it is hitting a case where
spaces are better. Or vim is doing it. Anyway your best bet is to use copy
mode.

**Comment by nicm** [comment]
What does this:

tmux info|grep Ms

Show from a shell inside tmux?

**Comment by mikehaertl** [comment]
> Oh I was wrong, tmux will already use BCE to clear when appropriate,

I can't confirm that. And I think I can rule out the terminal or Vim, because it works fine in both, as long as I don't use tmux.

```
$ tmux info|grep Ms
180: Ms: (string) \033]52;%p1%s;%p2%s\007
```

Still hoping for an answer on how to copy things into the system clipboard from tmux.

**Comment by nicm** [comment]
This should work, if you copy something then do:

tmux lsb

From the shell inside tmux, do you see what you copied?

**Comment by egmontkob** [comment]
>     180: Ms: (string) \033]52;%p1%s;%p2%s\007

gnome-terminal [doesn't support clipboard manipulation](https://bugzilla.gnome.org/show_bug.cgi?id=795774).

**Comment by egmontkob** [comment]
> your best bet is to use copy mode.

Or let Vim do the copying. It has the advantage that it can reliably preserve leading spaces vs. tabs, and doesn't chop off the end of overlong lines that don't fit in the current viewport.

**Comment by mikehaertl** [comment]
> Or let Vim do the copying.

Finally a solution that works. On my system I have to use <kbd>"</kbd> <kbd>+</kbd> <kbd>y</kbd> from inside Vim to copy the selection into the system clipboard. It does not work over SSH - but at least something.

**Comment by nicm** [comment]
OK thanks @egmontkob.

@mikehaertl sounds like you need to either change terminal, or possibly let vim do it. If vim can set the clipboard with OSC 52 (I don't know if it can) you will need to be running a newish tmux, and do this:

set -g set-clipboard on

In . tmux.conf (the default is "external" so tmux will set the clipboard but not let programs inside do it).

**Comment by nicm** [comment]
If vim is setting the clipboard by using X directly it will not work over  ssh unless you forward X and make sure DISPLAY is correct inside tmux before starting vim (and fix it up it you ssh in again).

**Comment by mikehaertl** [comment]
Thanks. While the solution works to some extend and while I understand your concerns from a maintainers perspective, I'm still inclinged to say that the user experience with tmux in this regard could be better.

Don't get me wrong: I appreciate all your help and what you did with tmux in general. But given the fact that copy & paste works perfect outside tmux and fails within, most users would clearly say: "That's a well known tmux bug and they refuse to fix it" - no matter what technical explanantion you can come up with.

And changing terminal or using pretty complex flows just to be able to copy text is not something you can expect from the average user. They expect that things "just work".

To sum it up: I still vote for having another look at the BCE implementation in tmux, maybe something can still be improved.

**Comment by mikehaertl** [comment]
> and I don't think you really know enough about it to be making this kind of comment

I stopped reading after this. I tried to explain, what an average user experiences. So you say, that this software is not for the average user or - even worse - I'm not even average but below average. I won't continue on this level.

**Comment by nicm** [comment]
I said nothing of the sort, but I suggest you should take your condescending and ill-informed rubbish elsewhere anyway.

**Comment by corbolais** [comment]
wowowow! guys!

Please take a deep breath. If we're going down that rabbit hole nothing helpful will come out of it.

Focus on the issue at hands.

I can understand someone saying, this needs to "just work"TM. I myself want it happen that way. 

The only thing is, what route can we take to reach this goal? 

One answer is: 
- use copy mode, 

another is: 

- implement BCE, copy mode is cumbersome, does not work reliably (see @egmontkob comment about gnome-terminal and clipboard manipulation) and is way to convoluted to use it w/o having to spell a charm.

So, where can we meet? @nicm already said so he would review a PR but won't implement it. 

Let's put it this way:

**Who wants to do the dirty work earning lots of karma points?** ;-)

cheers

**Comment by mckellyln** [comment]
I use vi (vim) and copy a selection within vim (ie not shift+mouse but vim's actual selection) to the clipboard and then can paste (I use bracketed-paste) that in the shell cmd-line or into any other app without trailing spaces.

**Comment by Sheco** [comment]
I have also bumped into this issue, where raw copy/paste from within vim, when TERM=screen or TERM=screen-256color insert trailing spaces.

Using TERM=xterm works fine and I don't see any artifacts, at least not with the color schemes I tested, so that's good enough for me for now, but I don't know if that's the best solution, I guess only time will tell.

I wish I could understand the internal intricacies of the implementation and knew how to patch this.

**Comment by wuxmedia** [comment]
Just in case anyone wanders in looking for some other fix (setting xterm seems to mostly work) 
Another workaround is to use the copy an area (hold ctrl while highlighting) thing on most terminals (xterm and xfce4-terminal so far) which crops whitespace and adds  newlines.

**Comment by eblau** [comment]
I ended up using copy mode and it works well in conjunction with mouse mode and the different recipes you can find online to pipe the copy mode selection to xclip to make it available on the X clipboard. There are no trailing spaces present in the copied selection using this method. Since bce support seems so terminal-specific and fragile, it seems best to use copy mode. That said, before switching to copy mode, I would zoom a pane in order to copy and it would be cool if that also worked reliably without copying trailing whitespace.
